streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java

```diff
@@ -93,21 +93,24 @@ static void registerStateStores(final Logger log,
         }
 
         final TaskId id = stateMgr.taskId();
-        if (!stateDirectory.lock(id)) {
+        if (!stateDirectory.canTryLock(id, System.currentTimeMillis())) {
+            log.trace("Task {} is still not allowed to retry acquiring the state directory lock", id);
```

Comment:
@aliehsaeedii @lucasbru I do not understand this branch.
If the task is not allowed to lock the state directory, the register method will just return as if the state directory were locked. That means, the task thinks it is initialized and the task is added to the state updater without the lock on the state directory. Since the task thinks it successfully registered the state store, it will never try to register the state store and acquire the lock on the state directory again. Thus, the task will be an illegal state.

Reply:
@cadonna @lucasbru Could `StateManagerUtil.registerStateStores()` return a boolean instead of void? This can solve the issue, otherwise, if we throw an exception in the case of `stateDirectory.canTryLock() = false,` what would be the benefit of having backoff-retry for the lock? However this is not a simple small change. I assume we need to change many methods then.

Reply:
It could be easier to move all the back-off logic to `TaskManager`. Before attempting to initialize in `addTaskToStateUpdater`, check `canTryLock`. If we catch a `LockException`, create the backoffRecord. Just an idea which maybe would not require changing many methods.

Reply:
I also wanted to propose something similar what Lucas wrote. I think that is simpler. But why not have a private method `canTryInitializeIfNeeded()` in `TaskManager` and create the backoff record when the `LockException lockException` is thrown?  The initialization is the actual call we want to backoff, right?

Reply:
Yes, I think we are talking about the same thing. All logic would be inside `TaskManager`.

Reply:
this [PR](https://github.com/apache/kafka/pull/17209) may solve the issue. Could you please review it?